### ISSUE TITLE
Remove reliance on package unsafe

### ISSUE THREAD

**Issue by redloaf** [jackc/puddle#36]
Since #35 merged, it seems the only usage of `unsafe` in this library is in log.go's `log2Int` function. That function is unexported and uses generics but is only called in [one place](https://github.com/jackc/puddle/blob/bd09d14bd4018b6d65a9d7770e2f3ddf8b00af1c/pool.go#L551). Go has a high-performance, table-driven stdlib function, [`bits.Len64`](https://pkg.go.dev/math/bits#Len64), that computes the minimum number of bits necessary to represent a number (which is `floor(log2(val))+1`). Therefore it seems `log2Int(val)` could just be rewritten as `bits.Len64(val) - 1`. With range checking and casting, that might look like this:

```
// log2Int returns log2 of an integer. This function panics if val <= 0.
func log2Int(val int) uint8 {
	if val <= 0 {
		panic("log2 of non-positive number does not exist")
	}
	return uint8(bits.Len64(uint64(val))-1)
}
```

Example: https://go.dev/play/p/yZMqe7X_vFx

**Comment by jackc** [comment]
Seems reasonable to me. But @jan-dubsky wrote this code originally. @jan-dubsky what do you think?

**Comment by jan-dubsky** [comment]
Yes, this is possible. I personally do not feel that references to `unsafe` are something wrong, but if somebody feels like that, we can do this change.

A nice consequence of this change is that whole `log.go` file can removed and the `bits.Len64()` can be used directly in `pool.go`. To be honest, I haven't known about `bits.Len64()` till now. If I knew it back then, I wouldn't probably write the `log.go` at all.

**Comment by jackc** [comment]
Okay. 👍 If someone wants to write the PR I think this is an easy merge.

**Comment by flowchartsman** [comment]
Here you go!

**Comment by flowchartsman** [comment]
Feel free to check out [e4eab01](https://github.com/jackc/puddle/pull/40/commits/e4eab01148653bb0400d00a0ce60bad8de718694) to verify that the original tests pass.
